lib/upgradedb/postgres.2021012700.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+/*
+ * LMS version 1.11-git
+ *
+ *  (C) Copyright 2001-2021 LMS Developers
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License Version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
+ *  USA.
+ *
+ */
+
+$this->BeginTrans();
+
+$this->Execute("
+    CREATE TABLE customercontactsproperties (
+    contactid   integer         NOT NULL
+        CONSTRAINT customercontactsproperties_contactid_fkey REFERENCES customercontacts (id) ON DELETE CASCADE ON UPDATE CASCADE,
+    name        varchar(255)    NOT NULL DEFAULT '',
+    value       varchar(255)    NOT NULL DEFAULT '',
+    CONSTRAINT customercontactsproperties_contactid_name_ukey UNIQUE (contactid, name)
```

Comment:
Gdy będziesz chciał usunąć właściwość to bęðziesz zmuszony podać jej nazwę, id kontaktu i wartość? Może wygodniej byłoby tu wprowadzic pole ID jako sekwencję z PRIMARY KEY()?

Reply:
Zastanawiałem się nad tym, ale poszedłem drogą podobną do tabeli 'customerconsents'. Zdecydowałem się na takie rozwiązanie bo:
- podczas edycji klienta wszystkie kontakty i tak są usuwane i tworzone na nowo więc zadziała ON DELETE CASCADE,
- jak usuwasz kontakt email dla klienta to i tak wszystkie kontakty są usuwane i tworzone na nowo więc znowu zadziała ON DELETE CASCADE,
- jak usuwasz tylko właściwość to w UI kasujesz zawartość nazwy i właściwości i zapisujesz, a kontakty i tak się wszystkie skasują i znowu utworzą więc znowu zadziała ON DELETE CASCADE

Nie widzę tutaj potrzeby wprowadzania sekwencji, bo nigdzie nie jest używana w zarządzaniu właściwościami kontaktu. No chyba, że dla wygody używania opcji usuwania z poziomu DB. To co wprowadzić PRIMARY KEY()?

Reply:
> * jak usuwasz tylko właściwość to w UI kasujesz zawartość nazwy i właściwości i zapisujesz, a kontakty i tak się wszystkie skasują i znowu utworzą więc znowu zadziała ON DELETE CASCADE

Tego nie rozumiem. Chodzi o to, że obecnie przy edycji klienta wszystkie kontakty kasują się i na nowo tworzą?
To trzeba poprawić, żeby tak nie było - żeby tylko przyrostowo pokazywało kontakty - to była robota na szybkiego, żeby jakoś tam działało.

Reply:
> Tego nie rozumiem. Chodzi o to, że obecnie przy edycji klienta wszystkie kontakty kasują się i na nowo tworzą?

Dokładnie tak. Dlatego wcześniej przesłałem poprawkę, żeby to przynajmniej w transakcję objąć.

> To trzeba poprawić, żeby tak nie było - żeby tylko przyrostowo pokazywało kontakty - to była robota na szybkiego, żeby jakoś tam działało.

Z tego co patrzyłem to tak jest od dawien dawna. ale może źle patrzyłem. To może w takim razie w tym PR wprowadzę pole ID jako sekwencję z PRIMARY KEY(), to później będzie podkład pod kolejną poprawkę do pokazywania kontaktów. Co Ty na to?

Reply:
Tak - oczywiście, że jest od dawien dawna - dobra - niech tak na razie zostanie ;-)